specs/chain/hs/src/Cardano/Spec/Chain/STS/Rule/Chain.hs

```diff
@@ -107,6 +108,50 @@ genesisHash :: Hash
 -- Not sure we need a concrete hash in the specs ...
 genesisHash = Crypto.Hash.hash ("" :: ByteString)
 
+--------------------------------------------------------------------------------
+-- Chain environment getters.
+--------------------------------------------------------------------------------
+
+-- | Getter for the protocol parameters contained in the environment.
+--
+-- We want to use the getter with lens operations such as '^.', however we do
+-- not want the getter to be able to modify the environment using lens
+-- operators such as '.~'
+--
+-- The type of '^.', which is just `view` with the arguments flipped, is:
+--
+-- > (^.) :: s -> Getting a s a -> a
+--
+-- Hence the type we gave to 'getPps'.
+--
+-- We could have used:
+--
+-- > Getter (Environment CHAIN) PParams
+--
+-- which is equivalent to:
+--
+-- > forall f . (Contravariant f, Functor f)
+-- > => (PParams -> f PParams) -> (Environment CHAIN) -> f (Environment CHAIN)
+--
+-- However @Contravariant f@ is a redundant constraint, and GHC will give a warning.
+--
+-- The same remark applies to the other getters defined in this module.
+--
+getPps :: Getting PParams (Environment CHAIN) PParams
+getPps = _3
+
+--------------------------------------------------------------------------------
+-- Chain state getters.
+--------------------------------------------------------------------------------
+
+-- | Getter for the epoch contained in the chain state.
+getEpoch :: Getting Epoch (State CHAIN) Epoch
```

Comment:
Why are these not lenses? Modifying the state is certainly something we expect to do.

Reply:
Because I've been playing with the use of tuples for representing states, so that they match the `LaTeX` spec. In this style, we don't use lenses, but pattern match in the tuple. However we should decide which style we prefer: records (as in delegation or `UTxO`, or tuples, as in the chain spec).

Sorry for introducing the inconsistency, but I think it is worth considering both options (and deciding on one :) ).

Reply:
Right, but even there, if you are bothering to introduce a getter, why not make it a lens?

Reply:
Only because that the only thing I need at the moment, and I didn't think anyone would want to manipulate the chain state outside the scope of this module.